certora/specs/Splits.spec

```diff
@@ -0,0 +1,467 @@
+
+    uint128 splittableBefore;   uint128 collectableBefore;
+    uint128 splittableAfter;    uint128 collectableAfter;
+    
+    splittableBefore = splittable(e, userId, assetId);
+    collectableBefore = collectable(e, userId, assetId);
+
+    collectableAmt, splitAmt = split(e, userId, assetId, true);
+
+    splittableAfter = splittable(e, userId, assetId);
+    collectableAfter = collectable(e, userId, assetId);
+
+    assert splittableBefore >= splittableAfter;
+    assert collectableBefore + collectableAmt == collectableAfter;
+    assert splittableBefore + collectableBefore >= splittableAfter + collectableAfter;
+}
+
+
+/// @notice Collects user's received already split funds
+/// After collection, no more collectable balance should be immediately available
+rule integrityOfCollect() {
+    env e; uint256 userId; uint256 assetId;
+    uint128 collectedAmt; uint128 collectableAfter;
+
+    collectedAmt = collect(e, userId, assetId);
+    collectableAfter = collectable(e, userId, assetId);
+
+    assert collectableAfter == 0;
+}
+
+
+/// @notice Calling the method collect() should never revert
+rule revertCharacteristicsOfCollect() {
+    env e; uint256 userId; uint256 assetId; uint128 collectedAmt;
+
+    require e.msg.sender != 0;  // safe assumption that prevents revert
+    require e.msg.value == 0;  // safe assumption that prevents revert
+
+    collectedAmt = collect@withrevert(e, userId, assetId);
+    assert !lastReverted;
+}
+
+
+/// @notice The method give() gives funds from the user to the receiver.
+/// @param amt The given amount
+/// After give of amt, the splittable of the receiver should increase exactly by amt
+rule correctnessOfGive() {
+    env e; uint256 userId; uint256 receiver; uint256 assetId; uint128 amt;
+    uint128 splittableBefore; uint128 splittableAfter;
+
+    splittableBefore = splittable(e, receiver, assetId);
+    give(e, userId, receiver, assetId, amt);
+    splittableAfter = splittable(e, receiver, assetId);
+
+    assert splittableAfter == splittableBefore + amt;
+}
+
+
+/// @notice The method give() gives funds from the user to the receiver.
+/// The splittable of any other user that is not receiver should not change
+rule splittableOfNonReceiverNotAffectedByGive() {
+    env e; uint256 userId; uint256 receiver; uint256 assetId; uint128 amt;
+    uint256 otherUser; uint128 splittableBefore; uint128 splittableAfter;
+
+    require otherUser != receiver;
+
+    splittableBefore = splittable(e, otherUser, assetId);
+    give(e, userId, receiver, assetId, amt);
+    splittableAfter = splittable(e, otherUser, assetId);
+
+    assert splittableAfter == splittableBefore;
+}
+
+
+/// @notice the method hashSplits() Calculates the hash of the list of splits receivers.
+/// We use boolean selector to operate on two different lists of splits receivers.
+/// We verify that two calculated hashes are the same only if they got exactly the same input
```

Comment:
This version of the rule doesn't check that now.

Reply:
If I am not wrong, from a logical point of view, the two statements:
1) two calculated hashes are the same only if they got exactly the same input
2) different inputs must generate different hashes
Are equivalent

and the assert is (for any index value):
assert (receiversHash1 == receiversHash2) => ((length1 == length2) && (userId1 == userId2) && (weight1 == weight2));